pkg/volumes/azure/client.go

```diff
@@ -0,0 +1,257 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package azure
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"strings"
+
+	"github.com/Azure/azure-sdk-for-go/services/compute/mgmt/2020-06-01/compute"
+	"github.com/Azure/azure-sdk-for-go/services/network/mgmt/2020-06-01/network"
+	"github.com/Azure/go-autorest/autorest/azure/auth"
+)
+
+type managedDisk struct {
+	ID string `json:"id"`
+}
+
+type dataDisk struct {
+	Name        string       `json:"name"`
+	Lun         string       `json:"lun"`
+	ManagedDisk *managedDisk `json:"managedDisk"`
+}
+
+type storageProfile struct {
+	DataDisks []*dataDisk `json:"dataDisks"`
+}
+
+type instanceComputeMetadata struct {
+	Name              string          `json:"name"`
+	ResourceGroupName string          `json:"resourceGroupName"`
+	VMScaleSetName    string          `json:"vmScaleSetName"`
+	SubscriptionID    string          `json:"subscriptionId"`
+	StorageProfile    *storageProfile `json:"storageProfile"`
+}
+
+type ipAddress struct {
+	PrivateIPAddress string `json:"privateIpAddress"`
+	PublicIPAddress  string `json:"publicIpAddress"`
+}
+
+type ipv4Interface struct {
+	IPAddresses []*ipAddress `json:"ipAddress"`
+}
+
+type networkInterface struct {
+	IPv4 *ipv4Interface `json:"ipv4"`
+}
+
+type instanceNetworkMetadata struct {
+	Interfaces []*networkInterface `json:"interface"`
+}
+
+type instanceMetadata struct {
+	Compute *instanceComputeMetadata `json:"compute"`
+	Network *instanceNetworkMetadata `json:"network"`
+}
+
+type client struct {
+	metadata         *instanceMetadata
+	vmssesClient     *compute.VirtualMachineScaleSetsClient
+	vmssVMsClient    *compute.VirtualMachineScaleSetVMsClient
+	disksClient      *compute.DisksClient
+	interfacesClient *network.InterfacesClient
+}
+
+func newClient() (*client, error) {
+	m, err := queryInstanceMetadata()
+	if err != nil {
+		return nil, fmt.Errorf("error querying instance metadata: %s", err)
+	}
+	if m.Compute.SubscriptionID == "" {
+		return nil, fmt.Errorf("empty subscription name")
+	}
+	if m.Compute.ResourceGroupName == "" {
+		return nil, fmt.Errorf("empty resource group name")
+	}
+
+	authorizer, err := auth.NewAuthorizerFromEnvironment()
+	if err != nil {
+		return nil, fmt.Errorf("error creating an authorizer: %s", err)
+	}
+
+	vmssesClient := compute.NewVirtualMachineScaleSetsClient(m.Compute.SubscriptionID)
+	vmssesClient.Authorizer = authorizer
+
+	vmssVMsClient := compute.NewVirtualMachineScaleSetVMsClient(m.Compute.SubscriptionID)
+	vmssVMsClient.Authorizer = authorizer
+
+	disksClient := compute.NewDisksClient(m.Compute.SubscriptionID)
+	disksClient.Authorizer = authorizer
+
+	interfacesClient := network.NewInterfacesClient(m.Compute.SubscriptionID)
+	interfacesClient.Authorizer = authorizer
+
+	return &client{
+		metadata:         m,
+		vmssesClient:     &vmssesClient,
```

Comment:
Nit: These fields names are hard to read at first.  It might be clearer to drop `client` and just go with `scaleSets`, `scaleSetVMs`, `disks` and `networkInterfaces`.

But OTOH I'm also new to the azure SDK, so this might just be because I'm not familiar with it all!

Reply:
Agreed! Let me fix with a separate PR.